packages/venia-ui/lib/components/Main/main.js

```diff
@@ -19,7 +19,11 @@ const Main = props => {
     return (
         <main className={rootClass}>
             <Header />
-            <div className={pageClass}>{children}</div>
+            <div className={pageClass}>
+                <div id="main-ep-before-children" />
```

Comment:
> We'll have to add them all over the place [and] that still won't satisfy every use case our clients want

Like @sirugh said, we're just exploring happy paths for adding content to a storefront. We would only put extension points in the safest places to add content.

> Seemingly innocuous changes (I'm thinking `pageClass` turns from `flex` to `grid` or something) could result in completely broken extensions that our versioning strategy is never going to be able to account for.

This is true. In a normal app where users write global CSS targeting any classname, we'd assume that after adding content to an extension point, they'd write a CSS rule targeting its parent to ensure it was laid out correctly. This use case gets tougher with scoped CSS, though, since a user would have to take over the parent (defeating the purpose of this happy path).

Reply:
> I'm fearful these extension points will get out of control.

Slight cringe while reading this and hearing Ron Howard say "They did." Magento always had bottomless XML nodes that render any child, and later added a few explicit "extension placeholder" nodes.

In full honesty, I always hated these because when extensions hook onto them it never seems to be "the right place" for that extension — just an inoffensive place to put anything. Factor in responsive design (they didn't) and you always-always had to manually inject them somewhere else.

I understand it breaks the plug-n-play goal of cloud+extensions but I would've been happier to just see extensions define an explicit hook for itself and tell you (the developer integrating it) to add that one-line-of-code where it makes sense. The "safest places to add content" just weren't useful in the real client work.

Reply:
> I understand it breaks the plug-n-play goal of cloud+extensions but I would've been happier to just see extensions define an explicit hook for itself and tell you (the developer integrating it) to add that one-line-of-code where it makes sense. The "safest places to add content" just weren't useful in the real client work.

I agree, and my personal experience as an agency developer is in line with yours: for the best results, "real client work" *should* consist of developers manually integrating changes, regardless of whether those changes incorporate third-party code. Extensions that effect a seamless change are pretty rare in reality, even if they're targeting hard extension points.

That said, the plug & play extension story happens to be important to a lot of potential PWA Studio users. The lack of automatic extension integration seems to be holding some people back from trying out PWA Studio, even though they would likely find that manual integration leads to better experiences for shoppers and developers alike. A basic extension system would probably help us ease more people into this new ecosystem.

---

Zooming out, though, any approach to extensions would result in one of two outcomes:

1. If extensions are limited to only a few explicit placeholders, they'll be too limited for most scenarios, resulting in developers having to manually integrate their customizations anyway.

2. If extensions are *not* limited to a few explicit places, extensions will be able to undermine most assumptions about the page structure, resulting in defects and a general lack of predictability.

The latter is a net negative that we want to avoid. But the former should be able to provide risk-free value, even if it's only a small amount. If we can find a way to do this risk-free, we'll do it; if we can't, we won't do it.